### PR TITLE
Clean Graph layouts and increase flexibility

### PR DESCRIPTION
**Pull request by Nikhil-42**
## Overview: What does this pull request change?
This pull request changes the way `_determine_graph_layout` functions. It allows for user-supplied layout functions to be used in both the `__init__` and the `change_layout` function. Each of the customized layouts provided by Manim has been moved to a separate function that adheres to the new `LayoutFunction` interface. I have done my best to ensure backward compatibility.

+ standardizes layout function interfaces
+ adds support for user-supplied graph layout functions
+ improves the specificity of type hints in graph.py

## Motivation and Explanation: Why and how do your changes improve the library?
This pull request standardizes the interface of the various supported graph layouts to make code more consistent and simplify the introduction of future layout options. It provides an option for users to supply their own custom layout function to programmatically arrange graphs using `change_layout` or the `__init__`. 

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by MrDiver** [comment]
Sooo it would've been a good idea to make an issue for this first, before creating a PR with a lot of changes discussing the applications and changes to the code. Because it seems like a change that will affect a lot of the behavior of the current codebase.

So if you could provide some examples on cases where this is useful and how exactly it improves the consistency that would make reviewing this a lot easier. Because discussing ideas is a lot easier than figuring out the functionality of written code and discussing about the interpretation of that.

I am all for improving consistency and making the usage of the manim easier, but still examples are needed!

**Comment by Nikhil-42** [comment]
@MrDiver The `Graph` Mobject provides the `Graph.change_layout` method for arranging nodes. The user can currently specify the graph layout in 2 different ways: by name as a `str`, or by `dict[Hashable, np.ndarray[3]]`. The `Graph._determine_layout` function is used by manim to 'parse' the provided layout parameter and return a dictionary that maps nodes to positions. If the user supplied a `dict`, it is simply returned. If the user supplied a `str` name (`"tree"`, `"spring"`, etc.) then there is some extra work that goes on.

Of the supported named layouts, most are implemented by the underlying graph library NetworkX. If this is the case then the appropriate function is invoked and the results are cleaned up a little bit before returning. The `"tree"` and `"partite"` layouts are special cases. The `"tree"` layout is not implemented by NetworkX so it was implemented ad hoc in the `graph.py` file. The `"partite"` layout is implemented by NetworkX but there are extra steps before and after calling to NetworkX. These cases are checked in a large if-else block. Also, the `"tree"` and `"partite"` layouts require additional arguments which are passed as optional keyword arguments to the `Graph.__init__` and `Graph.change_layout`.

This pull request standardizes the interface of these layout functions by creating the `LayoutFunction` type (only statically-checked). It also allows the user to supply any function conforming to the `LayoutFunction` type to allow for new layouts at runtime. I also converted the `"tree"` and `"partite"` layouts to conform with `LayoutFunction`.

I tried to make all public interfaces fully backwards compatible, but in the future we can look at deprecating the `root_vertex` and `partitions` parameters instead requiring these arguments to be passed through the `layout_config` parameter thus decoupling the `Graph` constructor from the layout implementation.

**Comment by Nikhil-42** [comment]
@MrDiver Please let me know if there's anything I should improve.

**Comment by MrDiver** [comment]
I think it would be nice to have an Example for each of the layouts in the docs! (Not animated)

**Review by MrDiver — COMMENTED** [review]
LGTM for now, after Example scenes this can be merged.

File: manim/mobject/graph.py
Comment (MrDiver) [inline manim/mobject/graph.py:193]: It might be useful to add a type in typing.py with `Literal` such that you can type the `str` arguments and get autocompletion for which arguments are available

**Review by MrDiver — DISMISSED** [review]
Is that possible for you at the moment ? Or should we just do that ?

**Comment by Nikhil-42** [comment]
Sorry, it’s finals week for me right now so I haven’t had much free time 💀. I can get it done by this weekend but if you have the bandwidth feel free to do it earlier.

**Comment by MrDiver** [comment]
Then just do your things! If we find the time we can do it! No stress!

**Comment by Nikhil-42** [comment]
@MrDiver would you mind reviewing this MR

**Review by behackl — APPROVED** [review]
This is a significant improvement over the current situation! Sorry that this had to wait so long to be reviewed.

I've tested this locally, read the implementation (coarsely), and inspected the documentation and added tests. This all seems fine to me, thank you very much!

File: tests/module/mobject/test_graph.py
Comment (JasonGrace2282) [inline tests/module/mobject/test_graph.py:120]:
```suggestion
        layout for layout in _layouts if layout not in {"tree", "partite"}
```

File: tests/module/mobject/test_graph.py
Comment (JasonGrace2282) [inline tests/module/mobject/test_graph.py:168]:
```suggestion
        layout for layout in _layouts if layout not in {"tree", "partite"}
```

File: manim/mobject/graph.py
Comment (JasonGrace2282) [inline manim/mobject/graph.py:16]: I would prefer this to be under an `if TYPE_CHECKING` condition

File: manim/mobject/graph.py
Comment (JasonGrace2282) [inline manim/mobject/graph.py:27]: I would prefer this to be under an `if TYPE_CHECKING` condition as well

File: manim/mobject/graph.py
Comment (JasonGrace2282) [inline manim/mobject/graph.py:30]:
```suggestion
NxGraph: TypeAlias = nx.classes.graph.Graph | nx.classes.digraph.DiGraph
```
I would also prefer this to be under an `if TYPE_CHECKING` condition. Note that you might need to add a `from __future__ import annotations`

File: manim/mobject/graph.py
Comment (JasonGrace2282) [inline manim/mobject/graph.py:260]:
```suggestion
        *args: Any,
        **kwargs: Any,
```

File: manim/mobject/graph.py
Comment (JasonGrace2282) [inline manim/mobject/graph.py:283]:
```suggestion
    **kwargs: Any,
```

**Review by JasonGrace2282 — CHANGES_REQUESTED** [review]
Didn't take a look at the implementation, but left some comments regarding the typings of `args`/`kwargs` (see PEP [484](https://peps.python.org/pep-0484/#arbitrary-argument-lists-and-default-argument-values))

Seeing as this is already merged, I can make a follow-up PR for this if needed. Sorry for the late review!

**Pull request merged**
